lib/rspec/mocks/proxy.rb

```diff
@@ -149,7 +156,35 @@ def has_negative_expectation?(message)
       # @private
       def record_message_received(message, *args, &block)
         @order_group.invoked SpecificMessage.new(object, message, args)
-        @messages_received << [message, args, block]
+        @messages_received << [message, ReceivedArgList.new(args), block]
+      end
+
+      class ReceivedArgList
+        attr_reader :args
+
+        def initialize(args)
+          @args = args
+          @original_hashes = args.map { |arg| [arg, hash_of(arg)] }
```

Comment:
`Array#hash` will return a `hash` based on its components, so I think you can get away with just doing `@original_hash = @args.hash` and then comparing that later on.